### PR TITLE
feat(opcua source): add initial OPCUA source

### PR DESCRIPTION
**Pull request by simonhoss**
Hey guys

This is my first try adding OPCUA as source to vector ecosystem. Its my first rust pr ever, so any feedback is appreciated!

According ticket: #17593

### CODE REVIEW

**Comment by jszwedko** [comment]
Apologies for the delay here @simonhoss ! We'll be reviewing this soon. Thanks for the contribution!

**Comment by dsmith3197** [comment]
Hi @simonhoss, thank you for the contribution!

Just an FYI, per our [docs](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#new-sources-sinks-and-transforms), we will begin with completing that checklist prior to reviewing the code. No need to write up answers to the checklist questions at this stage, I will inquire about anything in this PR's comment thread.

As for now, I'm curious about two things:

> If this would be integrated we would be able to collect metrics and logs direct from machines without any converter.

Could you elaborate on your current use case? What converter do you use? Is this being used in an industrial or consumer setting?

>  (From the checklist) Is the contributor committed to maintaining the integration if it is accepted?

Would you be willing to maintain this source if it is accepted?

**Comment by simonhoss** [comment]
> Could you elaborate on your current use case? What converter do you use? Is this being used in an industrial or consumer setting?

OPC/UA is an industrial protocol and used in machine communication. Our use case is, that we want to grab machine data and forward it to a broker or directly save it in a database, for example an Influx. As an info at the moment it only grabs metrics and more specific floating metrics. I just wanted to have a starting point.

> Would you be willing to maintain this source if it is accepted?

Sure. We have some more use case which we want to extend as well.

**Comment by dsmith3197** [comment]
@simonhoss I have some good news, we've received the green light to add this source to Vector, and I will begin reviewing the code shortly.

**Review by dsmith3197 — CHANGES_REQUESTED** [review]
Thank you again for the contribution and for taking the time to learn rust! I'm impressed.

I took a preliminary look through this and left some comments, primarily focused on the configuration. I need some more time to throughly review the `opcua_source()` and related functions.  As you probably encountered, the `opcua` crate doesn't play too nicely with vector's tokio runtime because the crate exposes a synchronous API but uses tokio internally. To be clear, I don't see anything that would prevent us from merging this, but just need some time to think through it.

**Comment by simonhoss** [comment]
Thanks a lot. I will work on beginning next week!

File: src/sources/opcua/mod.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    #[configurable(metadata(docs::examples = "demo_val"))]
+    pub metric_name: String,
+}
+
+impl GenerateConfig for OpcUaSourceConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            url: "opc.tcp://localhost:4855".to_string(),
+            application_uri: "urn:SimpleClient".to_string(),
+            product_uri: "urn:SimpleClient".to_string(),
```
[the remark this answers is not in the document]
Comment (simonhoss) [reply src/sources/opcua/mod.rs:93]: Based on the original client example, it must not:

https://github.com/locka99/opcua/blob/master/samples/simple-client/src/main.rs#L56

File: src/sources/opcua/mod.rs [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+/// Configuration for the `nats` source.
+#[configurable_component(source(
+"opcua",
+"Read metrics data from opcua."
+))]
+#[serde(deny_unknown_fields)]
+#[derive(Clone, Debug, Derivative)]
+#[configurable(metadata(status = "beta"))]
+pub struct OpcUaSourceConfig {
+    /// The OPC/UA URL to connect to.
```
[the remark this answers is not in the document]
Comment (simonhoss) [reply src/sources/opcua/mod.rs:49]: I corrected it

File: src/sources/opcua/mod.rs [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    /// The URL takes the form of `opc.tcp://server:port`.
+    #[configurable(metadata(docs::examples = "opc.tcp://localhost:4840"))]
+    url: String,
+
+    /// The application URI to use when connecting to the server.
+    #[configurable(metadata(docs::examples = "urn:example:client"))]
+    application_uri: String,
+
+    /// The product URI to use when connecting to the server.
+    #[configurable(metadata(docs::examples = "product_uri"))]
```
[the remark this answers is not in the document]
Comment (simonhoss) [reply src/sources/opcua/mod.rs:60]: This is more for identifying. https://reference.opcfoundation.org/Core/Part4/v104/docs/7.1

**Comment by simonhoss** [comment]
@dsmith3197 Sorry for my delay. Would you mind checking the PR again?

**Comment by dsmith3197** [comment]
Hey @simonhoss, my apologies for not getting back to this in a timely manner. Would you still be interested in pushing this forward if we review it?

**Pull request closed without merging**